### DISCUSSION TITLE
How can i get metadata from source?

### DISCUSSION THREAD

**Discussion by rrolla** [savonet/liquidsoap#2859] [Q&A] [answered]
Is there a possibility to get metadata from source?

```ruby
channel1Playlist = playlist("/mnt/music")

def testMetaData() =
    lastMetaData = channel1Playlist.last_metadata()
    print("HOP #{lastMetaData['artist']}")
end

thread.run(delay=10., fun() -> begin testMetaData() {} end )
```

This throw error
```ruby
print("HOP #{lastMetaData['artist']}")

Error 5: this value has type
 _? 
but it should be a subtype of
[_]
```

**Comment by vitoyucepi** [comment, marked as answer]
Hi @rrolla,
You can use 
- `null.get()`
  ```ruby
  def testMetaData() =
    lastMetaData = null.get(default = [], channel1Playlist.last_metadata())
    print("HOP #{lastMetaData['artist']}")
  end
  ```
- `??`
  ```ruby
  def testMetaData() =
    lastMetaData = channel1Playlist.last_metadata() ?? []
    print("HOP #{lastMetaData['artist']}")
  end
  ```

**Reply by rrolla** [reply]
hey, @vitoyucepi will try your solution. Before I tried with `null.defined` it looks like I was close.

**Reply by vitoyucepi** [reply]
> `last_metadata` (of type `() -> [string * string]?`): Return the last metadata from the source.

https://github.com/savonet/liquidsoap/blob/5aa7811b5b41b8f21d5e77b4eb7e96a76e90e1b1/src/libs/null.liq#L7-L15

**Reply by rrolla** [reply]
Big thanks @vitoyucepi it works :) I will use `channel1Playlist.last_metadata() ?? []` this looks more readable than `null.get` way
